src/util/expr_iterator.h

```diff
@@ -215,17 +227,51 @@ class const_depth_iteratort final:
 class depth_iteratort final:
   public depth_iterator_baset<depth_iteratort>
 {
+private:
+  std::function<exprt &()> mutate_root;
+
 public:
   /// Create an end iterator
   depth_iteratort()=default;
   /// Create iterator starting at the supplied node (root)
   /// All mutations of the child nodes will be reflected on this node
-  explicit depth_iteratort(exprt &expr):
-    depth_iterator_baset(expr) { }
+  explicit depth_iteratort(exprt &expr) : depth_iterator_baset(expr)
+  {
+  }
+  /// Create iterator starting at the supplied node (root)
+  /// All mutations of the child nodes will be reflected on this node
+  explicit depth_iteratort(
+    const exprt &expr,
+    std::function<exprt &()> mutate_root)
+    : depth_iterator_baset(expr), mutate_root(mutate_root)
+  {
+    PRECONDITION(mutate_root);
+  }
+
+  /// Convert iterator to non-const
+  /// \remarks You must pass a non-const reference to the root exprt that
+  /// this iterator was created from to validate that you have non-const
+  /// access to this expression
+  /// \param expr: a non-const reference to the root exprt that this iterator
+  /// was created from
+  /// \returns A non-const iterator pointing to the same element as this
+  /// iterator
+
   /// Obtain non-const reference to the pointed exprt instance
   /// Modifies root expression
   exprt &mutate()
-  { return depth_iterator_baset::mutate(); }
+  {
+    if(mutate_root)
+    {
+      exprt &expr = mutate_root();
```

Comment:
What would such a function look like that permits creating a copy without even taking a parameter. Presumably a member of a class that stores the original `exprt`. Now I am starting to wonder: why not just keep a `const` and a non-`const` reference in the iterator from the start? You can then use the `const` one to iterate without impacting sharing.

Reply:
I'll add more comments explaining the intended use-case.